src/lib.rs

```diff
@@ -85,70 +84,83 @@ pub fn select_coins_bnb<T: Utxo>(
+                utxo_solution_set.push(u.clone());
+            }
+
+            Some(utxo_solution_set)
+        }
+        None => None,
+    }
 }
 
 fn find_solution<T: Utxo>(
     target: u64,
     cost_of_change: u64,
-    utxo_pool: &mut [T],
-) -> Option<Vec<bool>> {
+    utxo_pool: &[T],
+) -> Option<Vec<u64>> {
     let utxo_sum = utxo_pool.iter().fold(0u64, |mut s, u| {
         s += u.get_value();
         s
     });
 
-    let utxo_pool_length = utxo_pool.len();
-    utxo_pool.sort_by_key(|u| Reverse(u.get_value()));
-
-    let mut curr_selection: Vec<bool> = vec![false; utxo_pool_length];
-    let mut best_selection = None;
-    let mut remainder = utxo_sum;
+    let utxo_values: Vec<u64> = utxo_pool.iter().map(|x| x.get_value()).collect();
 
     let lower_bound = target;
     let upper_bound = cost_of_change + lower_bound;
 
+    let mut best_sum = utxo_sum;
+    let mut best_set = None;
+
     if utxo_sum < lower_bound {
         return None;
     }
 
-    for m in 0..utxo_pool_length {
-        let mut curr_sum = 0;
-        let mut slice_remainder = remainder;
-
-        for n in m..utxo_pool_length {
-            if slice_remainder + curr_sum < lower_bound {
-                break;
-            }
+    let set_size = utxo_pool.len();
+
+    // If all utxo's in the pool create a match
+    // then record that as a solution.  Note that this
+    // needs to be done since the Combinations interface
+    // won't allow selecting a set that is the same size
+    // as the input set. (n choose k) where n can't be the
+    // same size as k.
+    if utxo_sum >= lower_bound && utxo_sum <= upper_bound {
+        let mut v = Vec::with_capacity(utxo_pool.len());
+        for u in utxo_pool {
+            v.push(u.get_value());
+        }
+        best_set = Some(v);
+    }
 
-            let utxo_value = utxo_pool[n].get_value();
-            curr_sum += utxo_value;
-            curr_selection[n] = true;
+    // Start with set size of 1 and then increment
+    // the set size by one until all possible combination
+    // sizes are exhausted.
+    for i in 1..set_size {
+        let candidate_sets: Vec<_> = Combinations::new(utxo_values.clone(), i).collect();
 
-            if curr_sum >= lower_bound {
-                if curr_sum <= upper_bound {
-                    best_selection = Some(curr_selection.clone());
-                }
+        for s in candidate_sets {
+            let sum: u64 = s.iter().sum();
 
-                curr_selection[n] = false;
-                curr_sum -= utxo_value;
+            if sum < lower_bound || sum > upper_bound {
+                continue;
             }
 
-            slice_remainder -= utxo_value;
+            if sum < best_sum {
+                best_sum = sum;
+                best_set = Some(s)
+            }
         }
-
-        remainder -= utxo_pool[m].get_value();
-        curr_selection[m] = false;
     }
 
-    best_selection
+    best_set
 }
```

Comment:
Better! AFAICT, this should test all possible input sets, however, I would expect this approach to become extremely slow starting somewhere in the range of 30–50 UTXOs simply due to actually generating the entire combination space.

The advantage of using a Branch and Bound algorithm is that it allows to exclude large parts of the exponentially-sized search tree when you recognize certain circumstances.